File: C++/deque/deque.cpp
#include "deque.hpp"

template <typename T>
Deque<T>::Deque(T arr[], int length) : Deque() {
	for (int i = 0; i < length; ++i) {
		addFirst(arr[i]);
	}
}

template <typename T>
Deque<T>::Deque(const Deque& deque) : Deque() {
	DequeNode<T>* current = deque.front;
	while (current != nullptr) {
		addLast(current->data);
	}
}

template <typename T>
Deque<T>::~Deque() {
	if (size_ == 0) {
		return;
	}
	
	DequeNode<T>* current = front;
	DequeNode<T>* temp = front->next;
	while (temp != nullptr) {
		delete current;
		current = temp;
		temp = temp->next;
	}
	delete current;
}

template <typename T>
int Deque<T>::size() {
	return size_;
}

template <typename T>
bool Deque<T>::empty() {
	return size_ == 0;
}


template <typename T>
T& Deque<T>::getFirst() {
	if (front == nullptr) {
		return nullptr;
	}
	return front->data;
}

template <typename T>
T& Deque<T>::getLast() {
	if (tail_ == nullptr) {
		return nullptr;
	}
	return tail_->data;
}

template <typename T>
void Deque<T>::setFirst(T& e) {
	if (size_ == 0) {
		addFirst(e);
	}
	else {
		front->data = e;
	}
}

template <typename T>
void Deque<T>::setLast(T& e) {
	if (size_ == 0) {
		addLast(e);
	}
	else {
		tail_->data = e;
	}
}

template <typename T>
void Deque<T>::addFirst(T& e) {
	DequeNode<T>* temp = new DequeNode<T>(e);
	if (size_ == 0) {
		front = temp;
		tail_ = temp;
	}
	else {
		temp->next = front;
		temp->previous = nullptr;
		front = temp;
	}
	++size_;
}

template <typename T>
void Deque<T>::addLast(T& e) {
	DequeNode<T>* temp = new DequeNode<T>(e);
	if (size_ == 0) {
		tail_ = temp;
		front = temp;
	}
	else {
		tail_->next = temp;
		temp->previous = tail_;
		tail_ = temp;
	}
	++size_;
}

template <typename T>
void Deque<T>::removeFirst() {
	if (front == tail_) {
		clear();
	}
	else {
		DequeNode<T>* current = front->next;
		delete front;
		front = current;
		front->previous = nullptr;
		--size_;
	}
}

template <typename T>
void Deque<T>::removeLast() {
	if (front == tail_) {
		clear();
	}
	else {
		DequeNode<T>* current = tail_->previous;
		delete tail_;
		tail_ = current;
		tail_->next = nullptr;
		--size_;
	}
}

template <typename T>
bool Deque<T>::contains(T& e) {
	DequeNode<T>* current = front;
	while (current != nullptr) {
		if (current->data == e) {
			return true;
		}
		current = current->next;
	}
	return false;
}

template <typename T>
int Deque<T>::index(T& e) {
	DequeNode<T>* current = front;
	for (int i = 0; current != nullptr; ++i) {
		if (current->data == e) {
			return i;
		}
		current = current->next;
	}
	return -1;
}

template <typename T>
void Deque<T>::clear() {
	if (size_ == 0) {
		return;
	}
	
	DequeNode<T>* current = front;
	DequeNode<T>* temp = front->next;
	while (temp != nullptr) {
		delete current;
		current = temp;
		temp = temp->next;
	}
	delete current;
	
	front = nullptr;
	tail_ = nullptr;
	size_ = 0;
}

File: Java/deque/deque.java
package deque;

public class Deque<E> {
	private class DequeNode<E> {
		private E data;
		private DequeNode<E> next;
		private DequeNode<E> previous;
		
		private DequeNode(E e) {
			data = e;
			next = null;
			previous = null;
		}
	}
	
	private DequeNode<E> front;
	private DequeNode<E> back;
	private int size;
	
	public Deque() {
		front = null;
		back = null;
		size = 0;
	}
	
	public Deque(E[] elems) {
		this();
		for (E e : elems) {
			addLast(e);
		}
	}
	
	public int size() {
		return size;
	}
	
	public boolean isEmpty() {
		return size == 0;
	}
	
	public E getFirst() {
		if (front == null) {
			return null;
		}
		return front.data;
	}
	
	public E getLast() {
		if (back == null) {
			return null;
		}
		return back.data;
	}
	
	public E setFirst(E e) {
		E data = getFirst()
		if (size == 0) {
			addFirst(e);
		}
		else {
			front.data = e;
		}
		return data;
	}
	
	public void setLast(E e) {
		E data = getLast();
		if (size == 0) {
			addLast(e);
		}
		else {
			back.data = e;
		}
		return data;
	}
	
	public void addFirst(E e) {
		if (size == 0) {
			front = new DequeNode<E>(e);
			back = front;
		}
		else {
			DequeNode<E> temp = new DequeNode<E>(e);
			temp.next = front;
			front.previous = temp;
			front = temp;
		}
		size++;
	}
	
	public void addLast(E e) {
		if (size == 0) {
			back = new DequeNode<E>(e);
			front = back;
		}
		else {
			DequeNode<E> temp = new DequeNode<E>(e);
			back.next = temp;
			temp.previous = back;
			back = temp;
		}
		size++;
	}
	
	public E removeFirst() {
		E data = getFirst();
		if (front == back) {
			clear();
		}
		else {
			front = front.next;
			front.previous = null;
			size--;
		}
		return data;
	}
	
	public E removeLast() {
		E data = getLast();
		if (front == back) {
			clear();
		}
		else {
			back = back.previous;
			back.next = null;
			size--;
		}
		return data;
	}
	
	public boolean contains(E e) {
		DequeNode<E> current = front;
		while (current != null) {
			if (current.data.equals(e)) {
				return true;
			}
			current = current.next;
		}
		return false;
	}
	
	public void clear() {
		front = null;
		back = null;
		size = 0;
	}
}
